server/schedule/region_splitter_test.go

```diff
@@ -0,0 +1,110 @@
+	"github.com/tikv/pd/server/core"
+)
+
+type mockSplitRegionsHandler struct {
+	// regionID -> splitKeys
+	groupKeysByRegion map[uint64][][]byte
+	// regionID -> startKey, endKey
+	regions map[uint64][2][]byte
+}
+
+func newMockSplitRegionsHandler() *mockSplitRegionsHandler {
+	return &mockSplitRegionsHandler{
+		groupKeysByRegion: map[uint64][][]byte{},
+		regions:           map[uint64][2][]byte{},
+	}
+}
+
+// SplitRegionByKeys mock SplitRegionsHandler
+func (m *mockSplitRegionsHandler) SplitRegionByKeys(region *core.RegionInfo, splitKeys [][]byte) error {
+	m.regions[region.GetID()] = [2][]byte{
+		region.GetStartKey(),
+		region.GetEndKey(),
+	}
+	m.groupKeysByRegion[region.GetID()] = splitKeys
+	return nil
+}
+
+// WatchRegionsByKeyRange mock SplitRegionsHandler
+func (m *mockSplitRegionsHandler) WatchRegionsByKeyRange(startKey, endKey []byte, timeout, watchInterval time.Duration) []uint64 {
+	for regionID, keyRange := range m.regions {
+		if bytes.Equal(startKey, keyRange[0]) && bytes.Equal(endKey, keyRange[1]) {
+			returned := []uint64{regionID}
+			for i := 0; i < len(m.groupKeysByRegion[regionID]); i++ {
+				returned = append(returned, returned[0]+uint64(i)+1000)
+			}
+			return returned
+		}
+	}
+	return nil
+}
+
+var _ = Suite(&testRegionSplitterSuite{})
+
+type testRegionSplitterSuite struct{}
+
+func (s *testRegionSplitterSuite) TestRegionSplitter(c *C) {
+	opt := config.NewTestOptions()
+	tc := mockcluster.NewCluster(opt)
+	handler := newMockSplitRegionsHandler()
+	tc.AddLeaderRegionWithRange(1, "eee", "hhh", 2, 3)
+	splitter := NewRegionSplitter(tc, handler)
+	newRegions := map[uint64]struct{}{}
+	// assert success
+	failureKeys := splitter.splitRegionsByKeys([][]byte{[]byte("fff"), []byte("ggg")}, newRegions)
+	c.Assert(len(failureKeys), Equals, 0)
+	c.Assert(len(newRegions), Equals, 3)
+
+	percentage, newRegionsID := splitter.SplitRegions([][]byte{[]byte("fff"), []byte("ggg")}, 1)
+	c.Assert(percentage, Equals, 100)
+	c.Assert(len(newRegionsID), Equals, 3)
+	// assert out of range
+	newRegions = map[uint64]struct{}{}
+	failureKeys = splitter.splitRegionsByKeys([][]byte{[]byte("aaa"), []byte("bbb")}, newRegions)
+	c.Assert(len(failureKeys), Equals, 2)
+	c.Assert(len(newRegions), Equals, 0)
+
+	percentage, newRegionsID = splitter.SplitRegions([][]byte{[]byte("aaa"), []byte("bbb")}, 1)
+	c.Assert(percentage, Equals, 0)
+	c.Assert(len(newRegionsID), Equals, 0)
+}
+
+func (s *testRegionSplitterSuite) TestGroupKeysByRegion(c *C) {
+	opt := config.NewTestOptions()
+	tc := mockcluster.NewCluster(opt)
+	handler := newMockSplitRegionsHandler()
+	tc.AddLeaderRegionWithRange(1, "aaa", "ccc", 2, 3)
+	tc.AddLeaderRegionWithRange(2, "ccc", "eee", 2, 3)
+	tc.AddLeaderRegionWithRange(3, "fff", "ggg", 2, 3)
+	splitter := NewRegionSplitter(tc, handler)
+	groupKeys, unprocessKeys := splitter.groupKeysByRegion([][]byte{
+		[]byte("bbb"),
+		[]byte("ddd"),
+		[]byte("fff"),
+		[]byte("zzz"),
+	})
+	c.Assert(len(groupKeys), Equals, 2)
```

Comment:
better to check in detail, not only length

Reply:
updated.